### PR TITLE
upgrade:Track the sub steps and nodes status in the progress library

### PR DESCRIPTION
**Pull request by jsuchome**
This is not extended to the upgrade of compute nodes, as relevant PR is not merged yet. I think it could be solved as a followup.

### CODE REVIEW

**Comment by jsuchome** [comment]
[(8) Use sub-steps during the last part of the upgrade, part I: controller nodes](https://trello.com/c/StuV7nHt/203-8-use-sub-steps-during-the-last-part-of-the-upgrade-part-i-controller-nodes)

File: crowbar_framework/app/models/api/node.rb [hunk 1]
```diff
@@ -136,9 +133,24 @@ def disable_pre_upgrade_attribute_for(name)
...
+        ip: @node.public_ip,
+        state: state,
+        role: role
+      )
+      if state == "upgraded"
+        progress = status.progress
+        remaining = progress[:remaining_nodes] - 1
+        upgraded = progress[:upgraded_nodes] + 1
+        ::Crowbar::UpgradeStatus.new.save_nodes(
+          upgraded, remaining < 0 ? 0 : remaining
```
Comment (MaximilianMeister) [inline crowbar_framework/app/models/api/node.rb]: can `remaining` be smaller than 0?

Comment (jsuchome) [inline crowbar_framework/app/models/api/node.rb]: Probably couldn't. If this were a last node, remaining would be 1 here, and lowered to 0, and the function would not be called next time.

I think.

Comment (MaximilianMeister) [inline crowbar_framework/app/models/api/node.rb]: it looks like that, that's why i was asking. but why do we need this check then?

Comment (jsuchome) [inline crowbar_framework/app/models/api/node.rb]: Probably because I did not think it though the first time ;-)

**Comment by jsuchome** [comment]
BTW, this is the status after upgrading controller nodes:

```
root@crowbar:/opt/dell # crowbarctl upgrade status
+--------------------------------+--------------------------------------+
| Status                         | Value                                |
+--------------------------------+--------------------------------------+
| current_step                   | nodes_db_dump                        |
| current_substep                | computes                             |
| current_node.name              | d52-54-77-77-77-01.vo6.cloud.suse.de |
| current_node.alias             | data1                                |
| current_node.ip                | 10.165.197.2                         |
| current_node.state             | upgraded                             |
| current_node.role              | controller                           |
| remaining_nodes                | 2                                    |
| upgraded_nodes                 | 2                                    |
| steps.upgrade_prechecks.status | passed                               |
| steps.upgrade_prepare.status   | passed                               |
| steps.admin_backup.status      | passed                               |
| steps.admin_repo_checks.status | passed                               |
| steps.admin_upgrade.status     | passed                               |
| steps.database.status          | passed                               |
| steps.nodes_repo_checks.status | passed                               |
| steps.nodes_services.status    | passed                               |
| steps.nodes_db_dump.status     | pending                              |
| steps.nodes_upgrade.status     | pending                              |
| steps.finished.status          | pending                              |
+--------------------------------+--------------------------------------+
```

**Comment by MaximilianMeister** [comment]
`save_error_state` is also still a bit misleading

**Comment by jsuchome** [comment]
> would you mind dropping the save_upgrade_state method as well and replace it with Rails.logger 
> statements?

@MaximilianMeister I'd like to address those in followup PR, the way it is now is not harmful.

**Pull request merged**